### PR TITLE
[#930] Add tenant-scoped command consumer

### PR DESCRIPTION
**Pull request by calohmn**
For #930:
This introduces a new command consumer on a tenant-scoped address.

This consumer handler will map the device id in the command message 'To' address to a corresponding 'via' device id (if applicable) and delegate the command either to a local command handler or to a consumer on another protocol adapter instance by way of the AMQP network.
As the existing command consumer address scheme isn't changed (yet), this change doesn't cause any incompatibilities.

Additional changes for getting this to work are in #1156 and #1157.

### CODE REVIEW

**Comment by calohmn** [comment]
Rebased and resolved merge conflicts.
I'll still refactor this a bit (aiming to replace the new `registrationClientFactory` param in the `CommandConsumerFactory`).

**Comment by BobClaerhout** [comment]
Any particular reason why this is not yet merged?
@calohmn: is the refactor big?

**Comment by calohmn** [comment]
I've done some refactoring:
- The mapping logic is now in a separate class (`GatewayMapper` and `GatewayMapperImpl`) that also implements the `ConnectionLifecycle` interface. An object of this class is exposed as a bean and autowired in `AbstractProtocolAdapterBase`. That allows for potentially exchanging the implementation.
- The `GatewayMapperImpl` uses the `RegistrationClientFactory` internally to query the device registry. The `HonoConnection` for that is a separate one compared to the connection for the existing `AbstractProtocolAdapterBase.commandConsumerFactory` instance.
- The command handling logic for mapping the device id and for delegating the command request back to the AMQP network is in separate Handler classes now (`GatewayMappingCommandHandler` and `DelegatingCommandHandler`). I'll add unit tests for these once we have agreed on the general class structure.
- I had to add a new `init(GatewayMapper)` method to the `CommandConsumerFactory` interface. I wanted to add the GatewayMapper instance in the constructor, but that constructor is called from the `AbstractAdapterConfig` class - the `AbstractProtocolAdapterBase.gatewayMapper` instance can't be used from there.

File: client/src/main/java/org/eclipse/hono/client/CommandConsumerFactory.java [hunk 1]
```diff
@@ -36,6 +36,16 @@ static CommandConsumerFactory create(final HonoConnection connection) {
...
 
+    /**
+     * Initializes the CommandConsumerFactory with the given GatewayMapper.
+     * <p>
+     * This needs to be called before calling any of the createCommandConsumer() methods.
+     *
+     * @param gatewayMapper The component mapping a command device id to the corresponding gateway device id.
+     * @throws NullPointerException if gatewayMapper is {@code null}.
+     */
+    void init(GatewayMapper gatewayMapper);
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/CommandConsumerFactory.java]: why can't we pass in the gateway mapper to the factory method and the `CommandConsumerFactoryImpl` constructor? You can then override `AbstractHonoClientFactory`'s *connect()* and *disconnect()* method to start/sop both connections, or wouldn't that work?

File: client/src/main/java/org/eclipse/hono/client/DelegatedCommandSender.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+
+package org.eclipse.hono.client;
+
+/**
+ * A sender to send delegated command messages.
```
[the remark this answers is not in the document]
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/DelegatedCommandSender.java]: That is a command message, first retrieved from the *tenant-scoped* consumer in the protocol adapter, and then *delegated* back to the AMQP message network so that it can be consumed by the *device-specific* consumer later.
I can extend the javadoc there.

File: client/src/main/java/org/eclipse/hono/client/GatewayMapper.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ *
+ */
+public interface GatewayMapper extends ConnectionLifecycle {
+
+    /**
+     * Determines the gateway device id for the given device id (if applicable).
+     * The value of the returned Future can be either
+     * <ul>
+     * <li>the gateway device id</li>
+     * <li>{@code null} if the device is configured to be accessed via a gateway but none is set</li>
```
[the remark this answers is not in the document]
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/GatewayMapper.java]: That is the case where the device is configured with a `via` property (ie. it is configured to be accessed via a gateway), but the `last-via` property is not set. 
That means that there hasn't been a message sent for this device yet.

File: client/src/main/java/org/eclipse/hono/client/GatewayMapper.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+
+package org.eclipse.hono.client;
+
+import io.vertx.core.Future;
+
+/**
+ * A mapping component that determines the corresponding gateway device id for a given device id.
```
[the remark this answers is not in the document]
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/GatewayMapper.java]: Ok, I would rephrase it like this:
`A component that maps a given device to the gateway through which data was last published for the given device.`

File: client/src/main/java/org/eclipse/hono/client/TenantScopedCommandConsumer.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.proton.ProtonMessageHandler;
+import io.vertx.proton.ProtonQoS;
+import io.vertx.proton.ProtonReceiver;
+
+/**
+ * A wrapper around an AMQP receiver link for consuming commands that will be delegated to the appropriated
+ * handler/consumer.
```
[the remark this answers is not in the document]
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/TenantScopedCommandConsumer.java]: Right, the info is rather vague here.
I would change the javadoc to:
````
A wrapper around an AMQP receiver link for consuming commands on a tenant-scoped address.
<p>
This class is used by the default {@link CommandConsumerFactory} implementation to receive commands from northbound applications.
````
What is done with the command message afterwards is out of the scope of this class.
I can extend the javadoc of the `CommandConsumerFactoryImpl` to describe the handling of command messages there.

File: core/src/main/java/org/eclipse/hono/util/HonoProtonHelper.java [hunk 6]
```diff
@@ -274,4 +278,24 @@ public static void closeAndFree(
...
     }
+
+    /**
+     * Returns a copy of the given message.
+     * This is a shallow copy of the <em>Message</em> object, except for the copied <em>Properties</em>.
+     *
+     * @param message The message to copy.
+     * @return The message copy.
+     */
+    public static Message getMessageCopyWithCopiedProperties(final Message message) {
```
[the remark this answers is not in the document]
Comment (calohmn) [reply core/src/main/java/org/eclipse/hono/util/HonoProtonHelper.java]: Ok, yes, will move it there.
Concerning the name: Well, we could use `getCopyWithCopiedProperties` (or `getCopyWithPropertiesCopy`). Not the prettiest, but I'd rather keep the `withCopiedProperties` part in there to emphasize what the method is doing.

File: client/src/main/java/org/eclipse/hono/client/CommandConsumerFactory.java (hunk 1 above)
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/CommandConsumerFactory.java]: I thought about this, but decided against it.
This would mean that all the `ConnectionLifecycle` members of `AbstractProtocolAdapterBase` are connected in the `AbstractProtocolAdapterBase.connectToService` method - **except** the `gatewayMapper` field, where the connect invocation is hidden in the `CommandConsumerFactoryImpl` class.
This fact plus needing to override the `AbstractHonoClientFactory` methods makes this solution IMHO inferior to just adding the `init(GatewayMapper)` method.

Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/CommandConsumerFactory.java]: Or do you mean this along with removing the GatewayMapper field from the `AbstractProtocolAdapterBase` class?
I guess that would make exchanging the GatewayMapper implementation in a protocol adapter different from the way it would be done for the other factories.

Comment (sophokles73) [reply client/src/main/java/org/eclipse/hono/client/CommandConsumerFactory.java]: I guess we do not need (direct) access to the gateway mapper from the protocol adapters anyway, do we? Why not make it an implementation detail of the `CommandConsumerFactory`?

File: core/src/main/java/org/eclipse/hono/util/HonoProtonHelper.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (sophokles73) [reply core/src/main/java/org/eclipse/hono/util/HonoProtonHelper.java]: why not simply *getShallowCopy* and let users read the JavaDoc? in the end that is why you wrote it, isn't it?

File: client/src/main/java/org/eclipse/hono/client/GatewayMapper.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sophokles73) [reply client/src/main/java/org/eclipse/hono/client/GatewayMapper.java]: Ah, can you somehow incorporate that into the JavaDoc?

File: core/src/main/java/org/eclipse/hono/util/HonoProtonHelper.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (calohmn) [reply core/src/main/java/org/eclipse/hono/util/HonoProtonHelper.java]: ok, I can live with that.

File: client/src/main/java/org/eclipse/hono/client/DelegatedCommandSender.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/DelegatedCommandSender.java]: See [commit](https://github.com/eclipse/hono/compare/08dfcff597695d2a40c589f4d25a36d52a5d470a..dcca29de3f945fbec1b912f5d6626174c4b5f341) below.

File: client/src/main/java/org/eclipse/hono/client/GatewayMapper.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/GatewayMapper.java]: See [commit](https://github.com/eclipse/hono/compare/08dfcff597695d2a40c589f4d25a36d52a5d470a..dcca29de3f945fbec1b912f5d6626174c4b5f341) below.

File: client/src/main/java/org/eclipse/hono/client/GatewayMapper.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/GatewayMapper.java]: See [commit](https://github.com/eclipse/hono/compare/08dfcff597695d2a40c589f4d25a36d52a5d470a..dcca29de3f945fbec1b912f5d6626174c4b5f341) below.

File: client/src/main/java/org/eclipse/hono/client/TenantScopedCommandConsumer.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/TenantScopedCommandConsumer.java]: See [commit](https://github.com/eclipse/hono/compare/08dfcff597695d2a40c589f4d25a36d52a5d470a..dcca29de3f945fbec1b912f5d6626174c4b5f341) below.

File: client/src/main/java/org/eclipse/hono/client/CommandConsumerFactory.java (hunk 1 above)
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/CommandConsumerFactory.java]: I've explored this solution a bit more and think we can go with it. I've amended the commit with the corresponding changes.

**Comment by calohmn** [comment]
Rebased to resolve merge conflict.

File: client/src/main/java/org/eclipse/hono/client/impl/DelegatedCommandSenderImpl.java [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+    private Future<ProtonDelivery> runSendAndWaitForOutcomeOnContext(final Message rawMessage, final Span span) {
+        return connection.executeOrRunOnContext(result -> {
+            if (sender.sendQueueFull()) {
+                final ServiceInvocationException e = new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE, "no credit available");
+                logError(span, e);
+                span.finish();
+                result.fail(e);
+            } else {
+                sendMessageAndWaitForOutcome(rawMessage, span).setHandler(result.completer());
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/DelegatedCommandSenderImpl.java]: `Future.completer()` will be removed in vert.x 4 so I wouldn't use it anymore ...

File: client/src/main/java/org/eclipse/hono/client/impl/DelegatingCommandHandler.java [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+    private static final Logger LOG = LoggerFactory.getLogger(DelegatingCommandHandler.class);
+
+    private final Function<String, Handler<CommandContext>> deviceSpecificCommandHandlerSupplier;
+    private final Function<String, Future<DelegatedCommandSender>> delegatedCommandSenderSupplier;
+
+    /**
+     * Creates a new DelegatingCommandHandler.
+     *
+     * @param deviceSpecificCommandHandlerSupplier Function to get an existing command handler for a given device.
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/DelegatingCommandHandler.java]: so, the function's parameter is the device ID?

File: client/src/main/java/org/eclipse/hono/client/impl/DelegatingCommandHandler.java [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+     * @param delegatedCommandSenderSupplier Supplies a sender for a given tenant to send the delegated command via the
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/DelegatingCommandHandler.java]: so, the function's parameter is the tenant ID?
can the returned Future fail? what if there is no sender for the given tenant?

File: client/src/main/java/org/eclipse/hono/client/impl/DelegatingCommandHandler.java [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+        delegatedCommandSender.setHandler(cmdSenderResult -> {
+            if (cmdSenderResult.succeeded()) {
+                final MessageSender sender = cmdSenderResult.result();
+                final String targetAddress = DelegatedCommandSenderImpl.getTargetAddress(tenantId, deviceId);
+                final String replyToAddress = command.isOneWay() ? null
+                        : String.format("%s/%s/%s",
+                                CommandConstants.COMMAND_ENDPOINT, tenantId, command.getReplyToId());
+                final Future<ProtonDelivery> protonDeliveryFuture = sender.sendAndWaitForOutcome(
+                        createDelegatedCommandMessage(command.getCommandMessage(), targetAddress, replyToAddress),
+                        commandContext.getTracingContext());
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/DelegatingCommandHandler.java]: Isn't this the specific command context handling that `DelegatedCommandSender` could/should encapsulate? Why don't we define a method `DelegatedCommandSender.send(CommandContext)`?

File: client/src/main/java/org/eclipse/hono/client/impl/DelegatingCommandHandler.java [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+                                new Exception("failed to send command message to downstream peer",
+                                        sendResult.cause()));
+                        commandContext.release();
+                    }
+                });
+            } else {
+                // failed to create sender
+                LOG.error("failed to create sender for sending command message to downstream peer", cmdSenderResult.cause());
+                TracingHelper.logError(commandContext.getCurrentSpan(),
+                        new Exception("failed to create sender for sending command message to downstream peer",
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/DelegatingCommandHandler.java]: I do not really like using `Exception` ...

File: client/src/main/java/org/eclipse/hono/client/impl/GatewayMappingCommandHandler.java [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+    public GatewayMappingCommandHandler(final GatewayMapper gatewayMapper, final Handler<CommandContext> nextCommandHandler) {
+        this.gatewayMapper = gatewayMapper;
+        this.nextCommandHandler = nextCommandHandler;
+    }
+
+    @Override
+    public void handle(final CommandContext originalCommandContext) {
+        final Command originalCommand = originalCommandContext.getCommand();
+        if (!originalCommand.isValid()) {
+            originalCommandContext.reject(new ErrorCondition(Constants.AMQP_BAD_REQUEST, "malformed command message"));
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/GatewayMappingCommandHandler.java]: do we still need to execute the rest of the method if the command is being rejected?
I think we also should re-issue some credit so that the application can send another command, don't you think?

File: client/src/main/java/org/eclipse/hono/client/impl/GatewayMappingCommandHandler.java [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+                        }
+                        final Command command = Command.from(originalCommand.getCommandMessage(), tenantId, lastViaDeviceId);
+                        commandContext = CommandContext.from(command, originalCommandContext.getDelivery(), originalCommandContext.getReceiver(), originalCommandContext.getCurrentSpan());
+                    }
+                    nextCommandHandler.handle(commandContext);
+                } else {
+                    // lastViaDeviceId is null - device hasn't connected yet
+                    LOG.error("last-via for device {} is not set", originalDeviceId);
+                    TracingHelper.logError(originalCommandContext.getCurrentSpan(), "last-via for device " + originalDeviceId + " is not set");
+                    originalCommandContext.release();
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/GatewayMappingCommandHandler.java]: I think we also should re-issue some credit so that the application can send another command, don't you think?

File: client/src/main/java/org/eclipse/hono/client/impl/GatewayMappingCommandHandler.java [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+                }
+            } else {
+                // failed to get last-via
+                LOG.error("error getting last-via for device {}", originalDeviceId, deviceIdFutureResult.cause());
+                TracingHelper.logError(originalCommandContext.getCurrentSpan(),
+                        new Exception("error getting last-via for device", deviceIdFutureResult.cause()));
+                originalCommandContext.release();
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/GatewayMappingCommandHandler.java]: I think we also should re-issue some credit so that the application can send another command, don't you think?

File: client/src/main/java/org/eclipse/hono/client/CommandConsumerFactory.java [hunk 15]
```diff
@@ -63,6 +64,7 @@ static CommandConsumerFactory create(final HonoConnection connection) {
      *         the cause of the failure</li>
      *         </ul>
      * @throws NullPointerException if any of tenant, device ID or command handler are {@code null}.
+     * @throws IllegalStateException if init() hasn't been called before.
```
[the remark this answers is not in the document]
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/CommandConsumerFactory.java]: No, I've removed it.

File: client/src/main/java/org/eclipse/hono/client/CommandConsumerFactory.java [hunk 16]
```diff
@@ -106,6 +108,7 @@ static CommandConsumerFactory create(final HonoConnection connection) {
      *         </ul>
      * @throws NullPointerException if tenant, device ID or command handler are {@code null}.
      * @throws IllegalArgumentException if the checkInterval is negative.
+     * @throws IllegalStateException if init() hasn't been called before.
```
[the remark this answers is not in the document]
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/CommandConsumerFactory.java]: No, I've removed it.

File: client/src/main/java/org/eclipse/hono/client/DelegatedCommandSender.java [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+
+package org.eclipse.hono.client;
+
+/**
+ * A sender to send command messages that are delegated to be processed by another command consumer.
+ * <p>
+ * This usually involves command messages first retrieved via a tenant-scoped consumer and then delegated back to the
+ * downstream peer so that they can be consumed by the device-specific consumer.
+ */
+public interface DelegatedCommandSender extends MessageSender {
```
[the remark this answers is not in the document]
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/DelegatedCommandSender.java]: Good point, I've added `sendCommandMessage(Command, SpanContext)`. Better structure now.

File: client/src/main/java/org/eclipse/hono/client/GatewayMapper.java [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+     * <ul>
+     * <li>the gateway device id</li>
+     * <li>{@code null} if the device is configured to be accessed via a gateway (i.e. one or more 'via' devices are
+     * set), but no 'last-via' device is set, meaning that no message has been sent yet for this device.</li>
+     * <li>the given device id if the device is not configured to be accessed via a gateway</li>
+     * </ul>
+     *
+     * @param tenantId The tenant identifier.
+     * @param deviceId The device identifier.
+     * @return Future containing the mapped gateway device id or {@code null}.
```
[the remark this answers is not in the document]
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/GatewayMapper.java]: I've adapted the javadoc.

File: client/src/main/java/org/eclipse/hono/client/impl/DelegatingCommandHandler.java (hunk 8 above)
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/impl/DelegatingCommandHandler.java]: I've adapted the javadoc.

File: client/src/main/java/org/eclipse/hono/client/impl/DelegatingCommandHandler.java (hunk 9 above)
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/impl/DelegatingCommandHandler.java]: I've adapted the javadoc.

File: client/src/main/java/org/eclipse/hono/client/impl/DelegatingCommandHandler.java (hunk 10 above)
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/impl/DelegatingCommandHandler.java]: Yes, added a method in the interface (see above).

File: client/src/main/java/org/eclipse/hono/client/impl/GatewayMappingCommandHandler.java (hunk 12 above)
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/impl/GatewayMappingCommandHandler.java]: Oh, thanks for spotting, missed the `return` statement there.
As for the credits: The `TenantScopedCommandConsumer` has its receiver link being created with the default (ie. non-empty) `prefetchSize`, so the credit handling should be automatic (in contrast to the `DeviceSpecificCommandConsumer` with `prefetchSize=0`).

File: client/src/main/java/org/eclipse/hono/client/impl/DelegatingCommandHandler.java (hunk 11 above)
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/impl/DelegatingCommandHandler.java]: I've removed creating an Exception there.

File: client/src/main/java/org/eclipse/hono/client/impl/GatewayMappingCommandHandler.java (hunk 13 above)
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/impl/GatewayMappingCommandHandler.java]: See above, credit-handling being automatic on that link.

File: client/src/main/java/org/eclipse/hono/client/impl/GatewayMappingCommandHandler.java (hunk 14 above)
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/impl/GatewayMappingCommandHandler.java]: See above, credit-handling being automatic on that link.

File: client/src/main/java/org/eclipse/hono/client/impl/DelegatedCommandSenderImpl.java (hunk 7 above)
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/impl/DelegatedCommandSenderImpl.java]: Ok, I've changed that.

**Comment by calohmn** [comment]
I've pushed the corresponding changes. 
I've also added a `TODO` in `DelegatedCommandSenderImpl.runSendAndWaitForOutcomeOnContext` concerning error-handling. To not further increase the size of this PR, I would fix this in a separate PR.

File: client/src/main/java/org/eclipse/hono/client/impl/DelegatingCommandHandler.java [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+                commandContext.modify(modified.getDeliveryFailed(), modified.getUndeliverableHere(), 0);
+                break;
+            case Released:
+                commandContext.release();
+                break;
+            default:
+                LOG.warn("got unexpected delivery outcome; remote state: {}", delegatedMsgDelivery.getRemoteState());
+                TracingHelper.logError(commandContext.getCurrentSpan(), "got unexpected delivery outcome; remote state: "
+                        + delegatedMsgDelivery.getRemoteState());
+                commandContext.getCurrentSpan().finish();
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/DelegatingCommandHandler.java]: My understanding is that at this point the application that sent the message is still expecting a disposition frame indicating the outcome of the transfer, isn't it? IMHO we need to invoke any of the `CommandContext`'s methods in order to finish the original message transfer here, i.e. even if we received an unexpected outcome, we should send a disposition with a supported outcome to the application, don't you think?.

It also feels strange that we finish the Span in this case but not in the other cases ...

File: client/src/main/java/org/eclipse/hono/client/impl/GatewayMappingCommandHandler.java [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+                    // lastViaDeviceId is null - device hasn't connected yet
+                    LOG.error("last-via for device {} is not set", originalDeviceId);
+                    TracingHelper.logError(originalCommandContext.getCurrentSpan(), "last-via for device " + originalDeviceId + " is not set");
+                    originalCommandContext.release();
+                }
+            } else {
+                // failed to get last-via
+                LOG.error("error getting last-via for device {}", originalDeviceId, deviceIdFutureResult.cause());
+                TracingHelper.logError(originalCommandContext.getCurrentSpan(),
+                        new Exception("error getting last-via for device", deviceIdFutureResult.cause()));
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/GatewayMappingCommandHandler.java]: `Exception`?

My understanding is that we run into this block also, if the GatewayMapper`s connection to the Device Registration service is lost, right?

File: client/src/main/java/org/eclipse/hono/client/impl/DelegatingCommandHandler.java (hunk 19 above)
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/impl/DelegatingCommandHandler.java]: Right, I've put a `commandContext.release()` in there. 
Besides the transaction related outcomes (`Declared` and `Transactional`), this block can only be reached with a `Received` delivery state, indicating partial transfer. I guess the latter case would be handled by the proton-j layer below and not be exposed here.

> It also feels strange that we finish the Span in this case but not in the other cases ...

In the other cases the span is finished by the `CommandContext#accept/reject/etc.` methods.

File: client/src/main/java/org/eclipse/hono/client/impl/GatewayMappingCommandHandler.java (hunk 20 above)
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/impl/GatewayMappingCommandHandler.java]: Yes, that would be a common cause for an error here. I've added a comment indicating this.

**Comment by sophokles73** [comment]
@dejanb do you also want to take a look?
